[로그인] 안돌 미션 제출합니다.

andole98 opened the pull request:
안녕하세요 화투님! 리뷰를 요청드리게 된 안돌입니다.
리뷰를 부탁드리며 3가지 여쭙고자 합니다.
1.
`Interceptor`를 활용하여 로그인 기능을 구현해 보았습니다.
로그인과 관련 없는, 예를들어 `ArticleController`에 대한 테스트는 어떻게 진행하는게 좋을지 고민을 많이 했습니다.
`WebTestClient`에 로그인 세션을 부여하는 것으로 해결하였는데, 더 좋은 방법이 있는지 궁금합니다.

2.
`UserController`에서 `UserService`에 로그인 처리를 위임할 때, 실패시 메시지를 띄워야 했습니다. Dto(Signup, Login)를 갖고 값을 넘겨받게 해주고 싶었는데요,
Dto의 필드가 **쓰일 경우가 있고 안 쓰일 경우가 있었습니다.**

성공시에는 userName이 필요하지만 실패시에는 userName이 아닌 message가 필요했습니다.
성공과 실패시 다른 타입으로 넘겨주면 메서드 리턴 타입이 달라지므로 안 될 것 같고,
컨트롤러에 에러처리 로직을 포함시키고 싶지 않았습니다. 그래서 일단은 Dto 필드를 비워둔 채 리턴하게 했습니다.
안 쓰이는 필드에 접근하지 않으면 NPE가 나지 않지만, 그래도 NPE의 가능성이 있는 것이 사실이라, 어떻게 해결하는게 좋을지 궁금합니다.

3.
도메인 유효성 검증은 어느 단계에서 처리하는게 좋을지 잘 모르겠습니다.
User의 이름, 비밀번호 등의 제약 조건은 레벨 1에서 배운대로 도메인에서 처리하는게 좋아 보였습니다.
구글 검색에서는 많은 자료가 Dto에서 처리하도록 권장하더군요...
유효성 검증은 도메인에서 처리하는게 좋을까요, Dto에서 처리하는게 좋을까요?

감사합니다.

ike5923 committed:
[게시글 생성/조회/수정/삭제] 아이크 미션 제출합니다. (#18)

* feat: add gradle dependency

* feat: get/post practice

* refactor: modify path

* fix: add responseBody annotation

* feat: click the button to move article-edit page

* test: modify uri

* feat: create and save article

* feat: write the article

* feat: look up the article

* feat: look up the article by id

* fix: find an article by id

* feat: update the article

* feat: delete the article

* feat: move to index from article-edit

* feat: setting devtools

* fix : UI bug

* refactor: remove duplicated html code

* refactor: remove duplicated index logo

* refactor: create ArticleService

* refactor: divide package

* refactor: divide package

* refactor: modify return value

* refactor: modify test cod

- fix bug
- extract duplicate fields

* feat: update README.md

* refactor: modify variable naming

* refactor : remove Article set method

* refactor: change from getting id to sending message to article

* refactor: ensure the list does not change when it is passed

* refactor: check convention

* refactor: check convention

* refactor: change path (absolute -> relative)

* refactor: move devtools property

* test: article update

* test: articleRepository save

* test: articleRepository findById

* test: articleRepository delete

* refactor: remove ArticleService

* refactor: test code convention

andole98 committed:
build: jpa 추가

andole98 committed:
feat: ArticleRepository 생성

1. JPA CrudRepository 상속.
2. 학습테스트

andole98 committed:
feat: MySQL 연동

1. application.properties 설정
2. 마크업 정리

andole98 committed:
test: ArticleControllerTest

1. ArticleRepository 이용하도록 변경
2. URL 수정

andole98 committed:
feat: gitignore

bin/ 추가

andole98 committed:
feat: User Entity 생성

1. UserRepository 준비
2. UserController 생성
3. user-list.html 마크업 수정

andole98 committed:
feat: UserController 생성

1. 엔드포인트 생성
2. 마크업 변경

andole98 committed:
feat: 로그인 / 회원가입 구현

andole98 committed:
feat: User 유효성 검사 추가

1. UserException 생성
2. 유저 이름, 비밀번호 정규식 매칭 로직 추가

andole98 committed:
refactor: User.isEqualTo() 메서드 생성

1. 유저가 입력한 로그인 정보와 매칭하는 메서드
2. equals hashcode

andole98 committed:
refactor: UserController 메서드 변경

1. Post /login /logout
2. HttpSession 사용

andole98 committed:
refactor: 템플릿 변경

1. signup.js : 두 번 입력한 패스워드가 서로 다를 경우 submit하지 않음.
2. 로그인 시 드롭다운 제어.

andole98 committed:
setting: 설정 변경

1. hibernate show sql

andole98 committed:
test: UserRepositoryTest 생성

andole98 committed:
feat: UserService 추가

1. 서비스와 컨트롤러간 데이터 이동을 위한 Dto 생성

andole98 committed:
feat: LoginInterceptor 생성

1. LoginInterceptorConfig 생성
2. 인터셉터 제외 URL 추가.

andole98 committed:
feat: ArticleService 생성

1. 서비스 레이어 추가
2. ArticleControllerTest 기본 인증 추가

andole98 committed:
feat: UserRepository 생성

1. findByEmail 메서드 추가

andole98 committed:
refactor: navbar 템플릿 조정

1. dropdown.html 삭제

andole98 committed:
feat: User 수정, 탈퇴 구현

1. Put, Delete 매핑
2. UserService 기능 추가
3. 테스트 작성

andole98 committed:
test: LoginInterceptorConfigTest 테스트

1. 인터셉터 테스트

andole98 committed:
test: 모든 테스트 확인

1. ddl-auto : create 로 변경
2. BeforeEach, AfterEach 추가

andole98 committed:
setting: 로그 설정

1. 테스트 패키지와 프로덕트 패키지 설정 분리
2. 템플릿 중복 제거

andole98 committed:
docs: README.md 추가

andole98 committed:
Revert "[게시글 생성/조회] 안돌 미션 제출합니다. (#3)"

This reverts commit db19f26192c46a1d0638f28236217bcf0267154b.

andole98 committed:
Merge branch 'andole87' into me

phs1116 reviewed (changes requested):
안녕하세요! 리뷰를 맡게된 리뷰어 화투입니다. :)
전체적으로 스프링 활용 매우 잘 해주신것 같아요 👍 
개발하시면서 상당히 많이 고민 하신 것 같습니다.
질문 사항에 대한 답변들, 그리고 몇가지 피드백 드렸는데요,
확인해보시고 궁금하신 점 있으시면 DM주세요 :)

andole98 committed:
refactor: User 필드를 객체로 래핑

1. UserName, UserPassword, UserEmail
2. Converter 생성

andole98 committed:
test: User 필드 객체 테스트

andole98 committed:
feat: LoginController 생성

1. LoginControllerAdvice 생성
2. LoginControllerTest

andole98 committed:
refactor: ArticleControllerAdvice 생성

1. ArticleException 체크
2. 404 view 리턴
3. ArticleController.setArticleModel() 추출

andole98 committed:
refactor: LoginInterceeptorConfig 패키지 이동

1. interceptor 패키지로 이동

andole98 committed:
refactor: Repository.save() 제거

andole98 committed:
refactor: web.support 패키지 생성

andole98 committed:
test: ArticleRepositoryTest 변경

1. Update, Delete 테스트 변경

andole98 committed:
feat: AuthedWebtestClient 생성

1. Basic Auth가 필요한 컨트롤러테스트의 SuperClass
2. ArticleControllerTest가 상속 후 인터셉터 우회

andole98 committed:
test: Service 테스트

1. UserServiceTest
2. ArticleServiceTest
3. LoginServiceTest

andole98 committed:
refactor: domain 패키지 분리

1. domain.User
2. domain.Article

andole98 committed:
test: LoginInterceptorTest 테스트

1. 인터셉터 동작 테스트
2. 인터셉터 제외 테스트

andole98 commented:
안녕하세요 화투님! 좋은 키워드를 많이 주셔서 너무 좋았습니다. 👍 
`ControllerAdvice`나 `ArgumentResolver`등 재미있게 익혔습니다.
다시 한번 리뷰 부탁드립니다!

phs1116 reviewed (changes requested):
안녕하세요! 리뷰어 화투입니다.
전체적인 피드백 반영 👍 
매우 구현 잘해주셨습니다.
질문주신 사항에 대한 답변들 및 간단한 피드백 추가했습니다.
확인 부탁드릴게요 :) 궁금하신 점 있으시면 DM주세요!

andole98 committed:
feat: LoginSessionManager 생성

andole98 committed:
refactor: Controller에 LoginSessionManager 사용

andole98 committed:
feat: 페이지네이션 구현

1. ArticlePageableRepository 생성. extends PagingAndSortingRepository
2. PageableArticleService 생성.
3. @DefaultPageable 세팅. sort = "id", size = 3, direction = DESC
4. @Deprecated 기존 레포지토리, 서비스

andole98 committed:
fix: PageableAticleService 수정

1. pageable 인자를 활용하도록 변경.

phs1116 reviewed (approved):
안녕하세요! 리뷰어 화투입니다.
피드백 반영 💯 
매우 잘 구현해주셨습니다.
이번 단계 머지하도록 하겠습니다.
지금 까지 개발하시느라 수고 많으셨습니다. :)
